File: android-example-app/app/src/main/java/ch/freshbits/pathshare/example/MainActivity.java
package ch.freshbits.pathshare.example;

import android.Manifest;
import android.app.Activity;
import android.app.AlarmManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;

import java.net.URL;
import java.util.Date;

import ch.freshbits.pathshare.sdk.Pathshare;
import ch.freshbits.pathshare.sdk.helper.InvitationResponseListener;
import ch.freshbits.pathshare.sdk.helper.PermissionRequester;
import ch.freshbits.pathshare.sdk.helper.ResponseListener;
import ch.freshbits.pathshare.sdk.helper.SessionResponseListener;
import ch.freshbits.pathshare.sdk.model.Destination;
import ch.freshbits.pathshare.sdk.model.Session;
import ch.freshbits.pathshare.sdk.model.UserType;

public class MainActivity extends Activity {
    private static final int TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS = 1;
    private static final String SESSION_PREFERENCES= "session";
    private static final String SESSION_ID_KEY = "session_id";

    Session mSession;
    Button mCreateButton;
    Button mJoinButton;
    Button mInviteButton;
    Button mLeaveButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initializeCreateButton();
        initializeJoinButton();
        initializeInviteButton();
        initializeLeaveButton();

        findSession();
    }

    private void initializeCreateButton() {
        setCreateButton(findViewById(R.id.create_session));
        getCreateButton().setEnabled(true);
        getCreateButton().setOnClickListener(view -> Pathshare.client().saveUser(
                "SDK User",
                "me@email.com",
                "+12345678901",
                UserType.TECHNICIAN,
                getResources().getDrawable(R.drawable.face, null),
                new ResponseListener() {
            @Override
            public void onSuccess() {
                Log.d("User", "Success");
                createSession();
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                Log.e("User", "Error: " + throwable.getLocalizedMessage());
            }
        }));
    }

    private void initializeJoinButton() {
        setJoinButton(findViewById(R.id.join_session));
        getJoinButton().setEnabled(false);
        getJoinButton().setOnClickListener(view -> joinSession());
    }

    private void initializeInviteButton() {
        setInviteButton(findViewById(R.id.invite_customer));
        getInviteButton().setEnabled(false);
        getInviteButton().setOnClickListener(view -> inviteCustomer());
    }

    private void initializeLeaveButton() {
        setLeaveButton(findViewById(R.id.leave_session));
        getLeaveButton().setEnabled(false);
        getLeaveButton().setOnClickListener(view -> leaveSession());
    }

    private void createSession() {
        try {
            Destination destination = new Destination.Builder()
                    .setIdentifier("w9823")
                    .setLatitude(37.7875694)
                    .setLongitude(-122.4112239)
                    .build();

            Date expirationDate = new Date(System.currentTimeMillis() + 3600000);

            mSession = new Session.Builder()
                    .setDestination(destination)
                    .setExpirationDate(expirationDate)
                    .setName("simple session")
                    .setSessionExpirationListener(() -> handleSessionExpiration())
                    .build();

            getSession().save(new ResponseListener() {
                @Override
                public void onSuccess() {
                    Log.d("Session", "Success");
                    saveSessionIdentifier();

                    getCreateButton().setEnabled(false);
                    getJoinButton().setEnabled(true);
                    getLeaveButton().setEnabled(true);
                }

                @Override
                public void onError(@NonNull Throwable throwable) {
                    Log.e("Session", "Error: " + throwable.getLocalizedMessage());
                }
            });

        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        }
    }

    private void joinSession() {
        if (getSession().isExpired()) { return; }

        if (hasLocationPermission() && hasAlarmPermission()) {
            performJoinSession();
        } else {
            if (!hasLocationPermission()) { requestLocationPermission(); }
            if (!hasAlarmPermission()) { requestAlarmPermission(); }
        }
    }

    private void inviteCustomer() {
        if (getSession().isExpired()) { return; }

        getSession().inviteUser(
                "Customer",
                UserType.MOTORIST,
                "customer@me.com",
                "+12345678901",
                true,
                new InvitationResponseListener() {
            @Override
            public void onSuccess(URL url) {
                Log.d("Invite", "Success");
                Log.d("URL", url.toString());
                getInviteButton().setEnabled(false);
                getLeaveButton().setEnabled(true);
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                Log.e("Invite", "Error: " + throwable.getLocalizedMessage());
            }
        });
    }

    private void requestLocationPermission() {
        PermissionRequester.requestPermission(this, TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS, Manifest.permission.ACCESS_FINE_LOCATION, R.string.permission_access_fine_location_rationale);
    }

    private boolean hasLocationPermission() {
        return PermissionRequester.hasPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);
    }

    private void requestAlarmPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            return;
        }

        Intent intent = new Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM);
        startActivity(intent);
    }

    private boolean hasAlarmPermission() {
        return Build.VERSION.SDK_INT < Build.VERSION_CODES.S || getAlarmManager().canScheduleExactAlarms();
    }

    private AlarmManager getAlarmManager() {
        return (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
    }

    private Context getContext() {
        return Pathshare.client().getContext();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                performJoinSession();
            } else {
                Toast.makeText(this, R.string.permission_access_fine_location_denied, Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void performJoinSession() {
        try {
            if (getSession().isUserJoined()) { return; }

            getSession().join(new ResponseListener() {
                @Override
                public void onSuccess() {
                    Log.d("Join", "Success");
                    getCreateButton().setEnabled(false);
                    getJoinButton().setEnabled(false);
                    getInviteButton().setEnabled(true);
                    getLeaveButton().setEnabled(true);
                }

                @Override
                public void onError(@NonNull Throwable throwable) {
                    Log.e("Join", "Error: " + throwable.getLocalizedMessage());
                }
            });
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void leaveSession() {
        if (getSession().isExpired()) { return; }

        try {
            getSession().leave(new ResponseListener() {
                @Override
                public void onSuccess() {
                    Log.d("Leave", "Success");
                    getCreateButton().setEnabled(true);
                    getJoinButton().setEnabled(false);
                    getInviteButton().setEnabled(false);
                    getLeaveButton().setEnabled(false);


                    deleteSessionIdentifier();
                }

                @Override
                public void onError(@NonNull Throwable throwable) {
                    Log.e("Leave", "Error: " + throwable.getLocalizedMessage());
                }
            });
        } catch (IllegalStateException e) {
            e.printStackTrace();
        }
    }

    private void saveSessionIdentifier() {
        SharedPreferences.Editor editor = getPreferences().edit();
        editor.putString(SESSION_ID_KEY, getSession().getIdentifier());
        editor.apply();
    }

    private void deleteSessionIdentifier() {
        getPreferences().edit().clear().apply();
    }

    private void handleSessionExpiration() {
        this.runOnUiThread(() -> {
            getInviteButton().setEnabled(false);
            getLeaveButton().setEnabled(false);
            getCreateButton().setEnabled(true);
            deleteSessionIdentifier();
            showToast("Session expired.");
        });
    }

    public void setSession(Session session) {
        mSession = session;
    }

    public Session getSession() {
        return mSession;
    }

    public void setCreateButton(Button createButton) {
        this.mCreateButton = createButton;
    }

    public void setJoinButton(Button joinButton) {
        this.mJoinButton = joinButton;
    }

    public void setInviteButton(Button inviteButton) {
        this.mInviteButton = inviteButton;
    }

    public void setLeaveButton(Button leaveButton) {
        this.mLeaveButton = leaveButton;
    }

    public Button getCreateButton() {
        return mCreateButton;
    }

    public Button getJoinButton() {
        return mJoinButton;
    }

    public Button getInviteButton() {
        return mInviteButton;
    }

    public Button getLeaveButton() {
        return mLeaveButton;
    }

    private void findSession() {
        String identifier = getPreferences().getString(SESSION_ID_KEY, null);

        if (identifier == null) { return; }

        Pathshare.client().findSession(identifier, new SessionResponseListener() {
            @Override
            public void onSuccess(Session session) {
                if (session == null || session.isExpired()) {
                    deleteSessionIdentifier();

                    getCreateButton().setEnabled(true);
                    getJoinButton().setEnabled(false);
                    getInviteButton().setEnabled(false);
                    getLeaveButton().setEnabled(false);

                } else {
                    Log.d("Session", "Name: " + session.getName());
                    session.setSessionExpirationListener(() -> handleSessionExpiration());

                    setSession(session);

                    getCreateButton().setEnabled(false);
                    getJoinButton().setEnabled(true);
                    getInviteButton().setEnabled(false);
                    getLeaveButton().setEnabled(true);
                }
            }

            @Override
            public void onError(@NonNull Throwable throwable) {
                showToast("Something went wrong.");
            }
        });
    }

    private void showToast(String message) {
        Toast toast = Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT);
        toast.show();
    }

    private SharedPreferences getPreferences() {
        return getApplicationContext().getSharedPreferences(SESSION_PREFERENCES, Context.MODE_PRIVATE);
    }
}

File: android-example-app/app-kotlin/src/main/java/ch/freshbits/pathshare/example/MainActivity.kt
package ch.freshbits.pathshare.example

import android.app.AlarmManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import ch.freshbits.pathshare.sdk.Pathshare
import ch.freshbits.pathshare.sdk.helper.*
import ch.freshbits.pathshare.sdk.model.Destination
import ch.freshbits.pathshare.sdk.model.Session
import ch.freshbits.pathshare.sdk.model.UserType
import java.net.URL
import java.util.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS = 1
        private const val SESSION_PREFERENCES = "session"
        private const val SESSION_ID_KEY = "session_id"
    }

    private lateinit var createButton: Button
    private lateinit var joinButton: Button
    private lateinit var inviteButton: Button
    private lateinit var leaveButton: Button

    private lateinit var session: Session

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initializeCreateButton()
        initializeJoinButton()
        initializeInviteButton()
        initializeLeaveButton()

        findSession()
    }

    private fun initializeCreateButton() {
        createButton = findViewById<View>(R.id.create_session) as Button
        createButton.isEnabled = true
        createButton.setOnClickListener {
            Pathshare.client().saveUser(
                    "SDK User Android",
                    "me@email.com",
                    "+12345678901",
                    UserType.TECHNICIAN,
                    resources.getDrawable(R.drawable.face, null),
                    object: ResponseListener {
                override fun onSuccess() {
                    Log.d("User", "Success")
                    createSession()
                }

                override fun onError(throwable: Throwable) {
                    Log.e("User", "Error: ${throwable.localizedMessage}")
                }
            })
        }
    }

    private fun initializeJoinButton() {
        joinButton = findViewById<View>(R.id.join_session) as Button
        joinButton.isEnabled = false
        joinButton.setOnClickListener { joinSession() }
    }

    private fun initializeInviteButton() {
        inviteButton = findViewById<View>(R.id.invite_customer) as Button
        inviteButton.isEnabled = false
        inviteButton.setOnClickListener { inviteCustomer() }
    }

    private fun initializeLeaveButton() {
        leaveButton = findViewById<View>(R.id.leave_session) as Button
        leaveButton.isEnabled = false
        leaveButton.setOnClickListener { leaveSession() }
    }

    private fun findSession() {
        val identifier = preferences().getString(SESSION_ID_KEY, null) ?: return

        Pathshare.client().findSession(identifier, object: SessionResponseListener {
            override fun onSuccess(session: Session?) = if (session == null || session.isExpired) {
                deleteSessionIdentifier()

                createButton.isEnabled = true
                joinButton.isEnabled = false
                inviteButton.isEnabled = false
                leaveButton.isEnabled = false
            } else {
                Log.d("Session", "Name: ${session.name}")
                session.sessionExpirationListener = SessionExpirationListener { handleSessionExpiration() }

                createButton.isEnabled = false
                joinButton.isEnabled = true
                inviteButton.isEnabled = false
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                showToast("Something went wrong")
            }
        })
    }

    private fun createSession() {
        val destination = Destination.Builder()
                .setIdentifier("w9823")
                .setLatitude(37.7875694)
                .setLongitude(-122.4112239)
                .build()

        val expirationDate = Date(System.currentTimeMillis() + 3600000)

        session = Session.Builder()
                .setDestination(destination)
                .setExpirationDate(expirationDate)
                .setName("simple session")
                .setSessionExpirationListener { handleSessionExpiration() }
                .build()

        session.save(object: ResponseListener {
            override fun onSuccess() {
                Log.d("Session", "Success")
                saveSessionIdentifier()

                createButton.isEnabled = false
                joinButton.isEnabled = true
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                Log.e("Session", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun joinSession() {
        if (session.isExpired) return

        if (hasLocationPermission() && hasAlarmPermission()) {
            performJoinSession()
        } else {
            if (!hasLocationPermission()) { requestLocationPermission() }
            if (!hasAlarmPermission()) { requestAlarmPermission() }
        }
    }

    private fun inviteCustomer() {
        if (session.isExpired) return

        session.inviteUser(
                "Customer",
                UserType.MOTORIST,
                "customer@me.com",
                "+12345678901",
                true,
                object: InvitationResponseListener {
            override fun onSuccess(url: URL?) {
                Log.d("Invite", "Success")
                Log.d("URL", url.toString())
                inviteButton.isEnabled = false
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                Log.e("Invite", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun leaveSession() {
        if (session.isExpired) return

        session.leave(object: ResponseListener {
            override fun onSuccess() {
                Log.d("Leave", "Success")
                createButton.isEnabled = true
                joinButton.isEnabled = false
                inviteButton.isEnabled = false
                leaveButton.isEnabled = false

                deleteSessionIdentifier()
            }

            override fun onError(throwable: Throwable) {
                Log.e("Leave", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun requestLocationPermission() {
        PermissionRequester.requestPermission(this, TAG_PERMISSIONS_REQUEST_LOCATION_ACCESS, android.Manifest.permission.ACCESS_FINE_LOCATION, R.string.permission_access_fine_location_rationale)
    }

    private fun performJoinSession() {
        if (session.isUserJoined) return

        session.join(object: ResponseListener {
            override fun onSuccess() {
                Log.d("Join", "Success")
                createButton.isEnabled = false
                joinButton.isEnabled = false
                inviteButton.isEnabled = true
                leaveButton.isEnabled = true
            }

            override fun onError(throwable: Throwable) {
                Log.e("Join", "Error: ${throwable.localizedMessage}")
            }
        })
    }

    private fun hasLocationPermission(): Boolean {
        return PermissionRequester.hasPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION)
    }

    private fun requestAlarmPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            return
        }
        val intent = Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM)
        startActivity(intent)
    }

    private fun hasAlarmPermission(): Boolean {
        return Build.VERSION.SDK_INT < Build.VERSION_CODES.S || getAlarmManager().canScheduleExactAlarms()
    }

    private fun getAlarmManager(): AlarmManager {
        return getContext().getSystemService(ALARM_SERVICE) as AlarmManager
    }

    private fun getContext(): Context {
        return Pathshare.client().context
    }

    private fun saveSessionIdentifier() {
        val editor = preferences().edit()
        editor.putString(SESSION_ID_KEY, session.identifier)
        editor.apply()
    }

    private fun handleSessionExpiration() {
        this.runOnUiThread {
            inviteButton.isEnabled = false
            leaveButton.isEnabled = false
            createButton.isEnabled = true
            deleteSessionIdentifier()
            showToast("Session expired")
        }
    }

    private fun showToast(message: String) {
        val toast = Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT)
        toast.show()
    }

    private fun deleteSessionIdentifier() {
        preferences().edit().clear().apply()
    }

    private fun preferences(): SharedPreferences {
        return applicationContext.getSharedPreferences(SESSION_PREFERENCES, MODE_PRIVATE)
    }
}
